File: src/surf/research/engine.ts
/**
 * Relay Surf — research engine (server-only).
 *
 * Runs the multi-step research loop for a session:
 *
 *   PLAN → SEARCH → OPEN → EXTRACT → ENOUGH? → synthesize
 *
 * with hard, configurable budgets (searches, pages, runtime, depth) checked at
 * every step, and early stoppage once enough independent evidence exists. Every
 * step is recorded as a human-readable `SurfAction` so the UI renders a live
 * activity stream — never raw logs. Pages that trip the prompt-injection scan
 * are quarantined and never enter the evidence pool.
 */
import type { SearchResult, Source, SurfPlan, SurfSession } from "../types";
import { resolveLimits, isBudgetExhausted } from "./limits";
import { planResearch } from "./planner";
import {
  createSurfSession,
  getSurfSession,
  getSessionStore,
  recordAction,
  finishAction,
  setSurfSessionStatus,
  syncEvidenceToSession,
} from "./session";
import { routeSearch, runDeepSearch } from "../search/router";
import { searchPlatform, adapterForUrl } from "../platforms";
import { attachScores, scoreSource } from "../evidence/score";
import { assertSafeUrl } from "../guard";

export interface StartResearchInput {
  objective: string;
  userId: string;
  workspaceId?: string;
  depth?: SurfPlan["depth"];
  maxSearches?: number;
  maxPages?: number;
  maxRuntimeMs?: number;
}

const DEEP_DEPTHS: SurfPlan["depth"][] = ["deep", "exhaustive"];

/**
 * Start a research run and return the session immediately. The loop runs in
 * the background (the Express process stays alive until it finishes); callers
 * poll the session via `getSurfSession` / the RPC layer.
 */
export async function startResearch(input: StartResearchInput): Promise<SurfSession> {
  const objective = input.objective.trim();
  if (!objective) throw new Error("A research objective is required.");

  const plan = await planResearch(objective, input.depth);
  const session = createSurfSession({
    userId: input.userId,
    workspaceId: input.workspaceId,
    objective,
    plan,
  });

  void runResearchLoop(session.id, {
    maxSearches: input.maxSearches,
    maxPages: input.maxPages,
    maxRuntimeMs: input.maxRuntimeMs,
  });

  return session;
}

interface LoopOptions {
  maxSearches?: number;
  maxPages?: number;
  maxRuntimeMs?: number;
}

async function runResearchLoop(sessionId: string, opts: LoopOptions): Promise<void> {
  const session = getSurfSession(sessionId);
  if (!session || !session.plan) {
    setSurfSessionStatus(sessionId, "failed", "Session or plan missing at loop start.");
    return;
  }

  const plan = session.plan;
  const limits = resolveLimits(plan.depth, opts);

  let searches = 0;
  let pages = 0;
  const startedAt = Date.now();

  try {
    const queries = [...plan.queries];
    let queueIndex = 0;

    while (queueIndex < queries.length) {
      const budget = isBudgetExhausted({ limits, searches, pages, depth: 0, startedAt });
      if (budget.exhausted) break;
      if (stopEarly(sessionId)) break;

      const query = queries[queueIndex++];
      searches++;

      const action = recordAction(sessionId, "search", `Searching: "${query.slice(0, 80)}"`);
      const deep = DEEP_DEPTHS.includes(plan.depth);
      let results: SearchResult[] = [];
      let provider = "";
      let note = "";

      try {
        if (deep) {
          const run = await runDeepSearch(query, { limit: 8 });
          results = run.results;
          provider = run.provider;
          note = run.note ?? "";
        } else {
          const run = await routeSearch({ query, intent: "general", options: { limit: 8 } });
          results = run.results;
          provider = run.provider;
          note = run.note ?? "";
        }
      } catch (err) {
        finishAction(action, "error", err instanceof Error ? err.message : String(err));
        continue;
      }

      if (results.length === 0) {
        finishAction(action, "error", note || "No results from any configured provider.");
        continue;
      }
      finishAction(action, "done", `${results.length} results via ${provider}`);

      for (const hit of results) {
        const budget = isBudgetExhausted({ limits, searches, pages, depth: 0, startedAt });
        if (budget.exhausted) break;
        if (stopEarly(sessionId)) break;
        pages++;

        await processHit(sessionId, hit);
      }

      // Re-query when the evidence pool is thin and the query queue is empty.
      if (queueIndex >= queries.length && !stopEarly(sessionId) && sessionEvidenceCount(sessionId) < 3 && searches < limits.maxSearches) {
        const followUp = followUpQuery(plan, queries);
        if (followUp) queries.push(followUp);
      }
    }

    // Platform-scoped pass: when the plan names a platform and general search
    // produced nothing, ask the platform adapter directly (honest about mode).
    if (sessionEvidenceCount(sessionId) === 0 && searches < limits.maxSearches) {
      const platform = (plan.sources as string[]).find((s) =>
        ["reddit", "github", "youtube", "x", "instagram", "linkedin"].includes(s),
      );
      if (platform) {
        const action = recordAction(sessionId, "platform_search", `Searching ${platform} directly…`);
        try {
          const res = await searchPlatform({ platform: platform as never, query: plan.objective, options: { limit: 5 } });
          if (res.results.length === 0) {
            finishAction(action, "error", res.note ?? `No accessible search on ${platform}.`);
          } else {
            finishAction(action, "done", `${res.results.length} results on ${platform}`);
            for (const hit of res.results.slice(0, limits.maxPages)) {
              if (pages >= limits.maxPages) break;
              pages++;
              await processHit(sessionId, hit);
            }
          }
        } catch (err) {
          finishAction(action, "error", err instanceof Error ? err.message : String(err));
        }
      }
    }

    syncEvidenceToSession(sessionId);
    setSurfSessionStatus(sessionId, "complete");
  } catch (err) {
    syncEvidenceToSession(sessionId);
    setSurfSessionStatus(sessionId, "failed", err instanceof Error ? err.message : String(err));
  }
}

/** Fetch, scan and extract evidence from a single result URL. */
async function processHit(sessionId: string, hit: SearchResult): Promise<void> {
  const store = getSessionStore(sessionId);
  if (!store) return;

  // Score the source from the hit (relevance against the objective).
  const [scored] = attachScores([hit], getSurfSession(sessionId)?.objective ?? "");
  const source: Source = { ...scored };
  store.trackSource(source);

  let url = hit.url;
  try {
    url = assertSafeUrl(hit.url);
  } catch {
    return;
  }

  const action = recordAction(sessionId, "open", `Opening: ${source.title.slice(0, 60)}`, { provider: hit.provider });
  let page;
  try {
    page = await adapterForUrl(url).extractPage(url);
  } catch {
    page = null;
  }

  if (!page) {
    finishAction(action, "error", "Page could not be loaded.");
    return;
  }
  finishAction(action, "done", `${page.text.length.toLocaleString()} chars extracted`);

  // Quarantine injection-suspect pages — never let them into the evidence pool.
  if (page.injection?.suspect) {
    recordAction(sessionId, "extract", `Quarantined: ${source.title.slice(0, 60)} (possible prompt injection)`);
    return;
  }

  extractEvidence(store, source, page.text, getSurfSession(sessionId)?.objective ?? "");
}

const STOPWORDS = new Set([
  "the", "and", "for", "with", "that", "this", "from", "have", "what", "when",
  "where", "about", "into", "over", "their", "they", "them", "there", "than",
  "then", "will", "would", "should", "could", "also", "were", "been", "being",
  "which", "while", "your", "our",
]);

function objectiveKeywords(objective: string): string[] {
  return objective
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter((word) => word.length > 3 && !STOPWORDS.has(word))
    .slice(0, 12);
}

/** Prefer sentences that mention the objective and carry concrete detail. */
function scoreSentence(sentence: string, keywords: string[], index: number): number {
  const lower = sentence.toLowerCase();
  let score = 0;
  for (const keyword of keywords) {
    if (lower.includes(keyword)) score += 1;
  }
  if (/\d/.test(sentence)) score += 0.5;
  if (sentence.length >= 120) score += 0.25;
  // Page leads usually state context first; prefer it, weakly.
  if (index < 4) score += 0.25;
  return score;
}

function normalizeText(text: string): string {
  return text.toLowerCase().replace(/[^a-z0-9\s]/g, "").replace(/\s+/g, " ").trim();
}

/** Drop sentences that merely restate evidence already captured (boilerplate). */
function isNearDuplicate(store: NonNullable<ReturnType<typeof getSessionStore>>, sentence: string): boolean {
  const norm = normalizeText(sentence);
  if (norm.length < 40) return true;
  return store.all().some((existing) => {
    const current = normalizeText(existing.excerpt);
    if (!current || current.length < 40) return false;
    return current.includes(norm) || norm.includes(current);
  });
}

/** Pull objective-relevant sentences from a page into the evidence store. */
function extractEvidence(
  store: NonNullable<ReturnType<typeof getSessionStore>>,
  source: Source,
  text: string,
  objective: string,
): void {
  const keywords = objectiveKeywords(objective);
  const candidates = text
    .split(/(?<=[.!?])\s+|\n+/)
    .map((s) => s.trim())
    .filter((s) => s.length >= 40 && s.length <= 320)
    .map((sentence, index) => ({ sentence, index }))
    .sort((a, b) => scoreSentence(b.sentence, keywords, b.index) - scoreSentence(a.sentence, keywords, a.index))
    .slice(0, 8);

  let added = 0;
  for (const { sentence } of candidates) {
    if (added >= 4) break;
    if (store.count() >= 12) break;
    if (isNearDuplicate(store, sentence)) continue;
    store.add({
      sourceUrl: source.url,
      sourceTitle: source.title,
      platform: source.platform,
      claim: sentence.slice(0, 200),
      excerpt: sentence,
      accessMode: source.accessMode,
      confidence: source.score?.overall ?? 0.6,
    });
    added++;
  }
}

/** Stop once we have a decent multi-source evidence base. */
function stopEarly(sessionId: string): boolean {
  const store = getSessionStore(sessionId);
  if (!store) return false;
  const distinctSources = new Set(store.all().map((e) => e.sourceUrl)).size;
  return store.count() >= 6 && distinctSources >= 3;
}

function sessionEvidenceCount(sessionId: string): number {
  return getSessionStore(sessionId)?.count() ?? 0;
}

function followUpQuery(plan: SurfPlan, queries: string[]): string | null {
  const platform = (plan.sources as string[]).find((s) =>
    ["reddit", "github", "youtube", "x"].includes(s),
  );
  if (!platform) return null;
  const variant = `${plan.objective} ${platform}`;
  return queries.includes(variant) ? null : variant;
}

/** Get a live snapshot of a session (actions + evidence + optional report). */
export function getResearch(id: string): SurfSession | undefined {
  const session = getSurfSession(id);
  if (session) syncEvidenceToSession(id);
  return session;
}

export { scoreSource };

File: app/research/engine.py
"""Research engine — multi-step research loop.

Runs: PLAN → SEARCH → OPEN → EXTRACT → ENOUGH? → synthesize

with hard, configurable budgets (searches, pages, runtime, depth) checked at
every step, and early stoppage once enough independent evidence exists. Pages
that trip the prompt-injection scan are quarantined.
"""

from __future__ import annotations

import asyncio
import re
import time
import uuid
from datetime import datetime, timezone
from typing import Optional

from ..core.types import (
    Evidence, SearchAccessMode, SearchResult, Source, SurfAction,
    SurfPlan, SurfSession, SurfSessionStatus,
)
from ..core.guard import assert_safe_url
from ..evidence.scoring import attach_scores, score_source, _objective_keywords, _score_relevance
from ..evidence.store import EvidenceStore
from ..search.router import route_search, run_deep_search
from ..extraction.page_data import extract_page_data
from ..core.http import relai_fetch

# ── In-memory session store ───────────────────────────────────────────

_sessions: dict[str, SurfSession] = {}
_stores: dict[str, EvidenceStore] = {}


def create_session(user_id: str, objective: str, plan: SurfPlan) -> SurfSession:
    session = SurfSession(
        id=uuid.uuid4().hex[:16],
        user_id=user_id,
        objective=objective,
        plan=plan,
    )
    _sessions[session.id] = session
    _stores[session.id] = EvidenceStore()
    return session


def get_session(session_id: str) -> Optional[SurfSession]:
    return _sessions.get(session_id)


def _get_store(session_id: str) -> Optional[EvidenceStore]:
    return _stores.get(session_id)


def record_action(session_id: str, kind: str, description: str) -> SurfAction:
    session = _sessions.get(session_id)
    if not session:
        return SurfAction(id="", kind=kind, description=description, status="error")
    action = SurfAction(
        id=uuid.uuid4().hex[:8],
        kind=kind,
        description=description,
        status="running",
    )
    session.actions.append(action)
    return action


def finish_action(action: SurfAction, status: str, detail: Optional[str] = None):
    action.status = status
    action.detail = detail
    action.ended_at = datetime.now(timezone.utc).isoformat()


# ── Plan Generation ───────────────────────────────────────────────────

async def plan_research(objective: str, depth: str = "standard") -> SurfPlan:
    """Generate a research plan with query variations."""
    keywords = _objective_keywords(objective)

    queries = [
        objective,
        f"{objective} overview",
        f"{objective} 2025 2026",
    ]
    if keywords:
        queries.append(" ".join(keywords[:5]))

    sources = ["web"]
    if any(w in objective.lower() for w in ["github", "code", "repo", "open source"]):
        sources.append("github")
    if any(w in objective.lower() for w in ["reddit", "community", "discussion"]):
        sources.append("reddit")

    return SurfPlan(
        objective=objective,
        queries=queries,
        sources=sources,
        required_information=[objective],
        verification_requirements=["cross-reference from multiple sources"],
        depth=depth,
    )


# ── Budget ────────────────────────────────────────────────────────────

def _resolve_limits(depth: str, overrides: Optional[dict] = None) -> dict:
    from ..core.types import DEPTH_LIMITS
    limits = DEPTH_LIMITS.get(depth, DEPTH_LIMITS["standard"]).copy()
    if overrides:
        for k, v in overrides.items():
            if v is not None:
                limits[k] = v
    return limits


def _is_budget_exhausted(limits: dict, searches: int, pages: int, started_at: float) -> bool:
    if searches >= limits.get("max_searches", 8):
        return True
    if pages >= limits.get("max_pages", 12):
        return True
    if (time.time() - started_at) * 1000 >= limits.get("max_runtime_ms", 120_000):
        return True
    return False


# ── Evidence Extraction ───────────────────────────────────────────────

STOPWORDS = frozenset([
    "the", "and", "for", "with", "that", "this", "from", "have", "what", "when",
    "where", "about", "into", "over", "their", "they", "them", "there", "than",
    "then", "will", "would", "should", "could", "also", "were", "been", "being",
    "which", "while", "your", "our",
])


def _score_sentence(sentence: str, keywords: list[str], index: int) -> float:
    score = sum(1 for kw in keywords if kw in sentence.lower())
    if re.search(r"\d", sentence):
        score += 0.5
    if len(sentence) >= 120:
        score += 0.25
    if index < 4:
        score += 0.25
    return score


def extract_evidence(store: EvidenceStore, source: Source, text: str, objective: str):
    """Pull objective-relevant sentences from a page into the evidence store."""
    keywords = _objective_keywords(objective)
    sentences = re.split(r"(?<=[.!?])\s+|\n+", text)
    candidates = [
        (s.strip(), i) for i, s in enumerate(sentences)
        if 40 <= len(s.strip()) <= 320
    ]
    candidates.sort(key=lambda x: _score_sentence(x[0], keywords, x[1]), reverse=True)

    added = 0
    for sentence, _ in candidates[:8]:
        if added >= 4:
            break
        if store.count() >= 12:
            break
        store.add(Evidence(
            source_url=source.url,
            source_title=source.title,
            platform=None,
            claim=sentence[:200],
            excerpt=sentence,
            access_mode=SearchAccessMode.PUBLIC,
            confidence=source.score.overall if source.score else 0.6,
        ))
        added += 1


def _should_stop_early(store: EvidenceStore) -> bool:
    """Stop once we have decent multi-source evidence."""
    return store.count() >= 6 and store.distinct_sources() >= 3


# ── Hit Processing ────────────────────────────────────────────────────

async def process_hit(session_id: str, hit: SearchResult):
    """Fetch, scan and extract evidence from a single result URL."""
    store = _get_store(session_id)
    session = _sessions.get(session_id)
    if not store or not session:
        return

    source = Source(
        title=hit.title, url=hit.url, access_mode=hit.access_mode,
        retrieved_at=time.strftime("%Y-%m-%dT%H:%M:%SZ"),
    )
    score = score_source(source, session.objective)
    source.score = score
    store.track_source(source)

    try:
        url = assert_safe_url(hit.url)
    except ValueError:
        return

    action = record_action(session_id, "open", f"Opening: {hit.title[:60]}")

    try:
        result = await relai_fetch(url, timeout_ms=15_000, retries=2)
        if not result.ok:
            finish_action(action, "error", "Page could not be loaded")
            return

        is_json = "json" in result.content_type
        page = extract_page_data(
            html=None if is_json else result.text,
            json_str=result.text if is_json else None,
            url=result.url,
            status=result.status,
            content_type=result.content_type,
            max_chars=20_000,
            fetched_at=result.fetched_at,
        )

        finish_action(action, "done", f"{len(page.text):,} chars extracted")

        # Quarantine injection-suspect pages
        if page.injection and page.injection.suspect:
            record_action(session_id, "extract", f"Quarantined: {hit.title[:60]} (possible prompt injection)")
            return

        extract_evidence(store, source, page.text, session.objective)
    except Exception as e:
        finish_action(action, "error", str(e)[:200])


# ── Research Loop ─────────────────────────────────────────────────────

async def _run_research_loop(session_id: str, opts: dict):
    session = _sessions.get(session_id)
    if not session or not session.plan:
        if session:
            session.status = SurfSessionStatus.FAILED
            session.error = "Session or plan missing at loop start"
        return

    plan = session.plan
    limits = _resolve_limits(plan.depth, opts)
    searches = 0
    pages = 0
    started_at = time.time()

    try:
        queries = list(plan.queries)
        qi = 0

        while qi < len(queries):
            if _is_budget_exhausted(limits, searches, pages, started_at):
                break
            if _should_stop_early(_get_store(session_id)):
                break

            query = queries[qi]
            qi += 1
            searches += 1

            action = record_action(session_id, "search", f'Searching: "{query[:80]}"')

            try:
                if plan.depth in ("deep", "exhaustive"):
                    result = await run_deep_search(query)
                else:
                    result = await route_search(query, intent="general")

                hits = result.results
                provider = result.provider
                note = result.note or ""
            except Exception as e:
                finish_action(action, "error", str(e)[:200])
                continue

            if not hits:
                finish_action(action, "error", note or "No results")
                continue

            finish_action(action, "done", f"{len(hits)} results via {provider}")

            for hit in hits:
                if _is_budget_exhausted(limits, searches, pages, started_at):
                    break
                if _should_stop_early(_get_store(session_id)):
                    break
                pages += 1
                # Convert dict hits to SearchResult
                sr = SearchResult(
                    url=hit.get("url", ""),
                    title=hit.get("title", ""),
                    snippet=hit.get("snippet", ""),
                    provider=hit.get("provider", "unknown"),
                    access_mode=SearchAccessMode.PUBLIC,
                )
                await process_hit(session_id, sr)

            # Re-query when evidence is thin
            if qi >= len(queries) and not _should_stop_early(_get_store(session_id)):
                store = _get_store(session_id)
                if store and store.count() < 3 and searches < limits["max_searches"]:
                    follow_up = f"{plan.objective} site:reddit.com OR site:github.com"
                    if follow_up not in queries:
                        queries.append(follow_up)

        sync_evidence(session_id)
        session.status = SurfSessionStatus.COMPLETE
        session.completed_at = datetime.now(timezone.utc).isoformat()

    except Exception as e:
        sync_evidence(session_id)
        session.status = SurfSessionStatus.FAILED
        session.error = str(e)[:500]


def sync_evidence(session_id: str):
    """Sync evidence store to session."""
    store = _get_store(session_id)
    session = _sessions.get(session_id)
    if store and session:
        session.evidence = store.all()
        session.sources = [
            Source(
                title=s.title, url=s.url, access_mode="public",
                retrieved_at=s.retrieved_at, score=s.score,
            )
            for s in store.sources()
        ]


# ── Public API ────────────────────────────────────────────────────────

async def start_research(
    objective: str,
    user_id: str = "anonymous",
    depth: str = "standard",
    workspace_id: Optional[str] = None,
    max_searches: Optional[int] = None,
    max_pages: Optional[int] = None,
    max_runtime_ms: Optional[int] = None,
) -> SurfSession:
    """Start a research run. Returns session immediately; loop runs in background."""
    if not objective.strip():
        raise ValueError("A research objective is required")

    plan = await plan_research(objective, depth)
    session = create_session(user_id, objective.strip(), plan)
    if workspace_id:
        session.workspace_id = workspace_id

    # Run in background
    opts = {}
    if max_searches is not None:
        opts["max_searches"] = max_searches
    if max_pages is not None:
        opts["max_pages"] = max_pages
    if max_runtime_ms is not None:
        opts["max_runtime_ms"] = max_runtime_ms

    asyncio.create_task(_run_research_loop(session.id, opts))
    return session


def get_research(session_id: str) -> Optional[SurfSession]:
    """Get a live snapshot of a research session."""
    session = _sessions.get(session_id)
    if session:
        sync_evidence(session_id)
    return session
